graylog2-server/src/main/java/org/graylog/plugins/views/search/rest/QueryValidationResource.java

```diff
@@ -0,0 +1,145 @@
+import org.graylog2.plugin.rest.PluginRestResource;
+import org.graylog2.shared.rest.resources.RestResource;
+import org.graylog2.shared.security.RestPermissions;
+
+import javax.inject.Inject;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+
+@RequiresAuthentication
+@Api(value = "Search/Validation")
+@Path("/search/validate")
+public class QueryValidationResource extends RestResource implements PluginRestResource {
+
+    private final QueryValidationService queryValidationService;
+    private final PermittedStreams permittedStreams;
+
+    @Inject
+    public QueryValidationResource(QueryValidationService queryValidationService, PermittedStreams permittedStreams) {
+        this.queryValidationService = queryValidationService;
+        this.permittedStreams = permittedStreams;
+    }
+
+    @POST
+    @Produces(MediaType.APPLICATION_JSON)
+    @ApiOperation("Validate a search query")
+    @NoAuditEvent("Only validating query structure, not changing any data")
+    public ValidationResponseDTO validateQuery(@ApiParam(name = "validationRequest") ValidationRequestDTO validationRequest) {
+
+        final ValidationRequest.Builder q = ValidationRequest.Builder.builder()
+                .query(validationRequest.query())
+                .timerange(Optional.ofNullable(validationRequest.timerange()).orElse(defaultTimeRange()))
+                .streams(adaptStreams(validationRequest.streams()))
+                .parameters(resolveParameters(validationRequest));
+
+        validationRequest.filter().ifPresent(q::filter);
+
+        final ValidationResponse response = queryValidationService.validate(q.build());
+        return ValidationResponseDTO.create(toStatus(response.status()), toExplanations(response));
+    }
+
+    private ImmutableSet<Parameter> resolveParameters(ValidationRequestDTO validationRequest) {
+        return validationRequest.parameters().stream()
+                .map(param -> param.applyBindings(validationRequest.parameterBindings()))
+                .collect(toImmutableSet());
+    }
+
+    private ValidationStatusDTO toStatus(ValidationStatus status) {
+        final ValidationStatusDTO statusDTO;
+        switch (status) {
+            case WARNING:
+                statusDTO = ValidationStatusDTO.WARNING;
+                break;
+            case ERROR:
+                statusDTO = ValidationStatusDTO.ERROR;
+                break;
+            default:
+                statusDTO = ValidationStatusDTO.OK;
+        }
+        return statusDTO;
+    }
+
+    private List<ValidationMessageDTO> toExplanations(ValidationResponse response) {
+        if (ValidationStatus.OK.equals(response.status())) {
```

Comment:
I am not sure if this is solid thinking in the conversion method: If there are explanations in the internal representation, I would always convert them here. If there are none, it's empty - unrelated to the status.

Reply:
good catch, thanks, it's a relict from the ES validation implementation, where the explanation was usually a parsed lucene query when everything was OK. So we needed to filter these out, because they were not anyhow relevant. Removing the IF now.